src/musicbrainz/chromaprinter.cpp

```diff
@@ -1,83 +1,106 @@
-    unsigned int read = pSoundSource->read(numFingerprintSamples, pData);
+    QString calcFingerprint(const Mixxx::AudioSourcePointer& pAudioSource) {
 
-    if (read!=numFingerprintSamples) {
-        qDebug() << "oh that's embarrasing I couldn't read the track";
-        return QString();
-    }
-    qDebug("reading file took: %d ms" , timerReadingFile.elapsed());
-
-    ChromaprintContext* ctx = chromaprint_new(CHROMAPRINT_ALGORITHM_DEFAULT);
-    // we have 2 channels in mixxx always
-    chromaprint_start(ctx, pSoundSource->getSampleRate(), 2);
-
-    QTime timerGeneratingFingerPrint;
-    timerGeneratingFingerPrint.start();
-    int success = chromaprint_feed(ctx, pData, numFingerprintSamples);
-    if (!success) {
-        qDebug() << "could not generate fingerprint";
-        delete [] pData;
-        return QString();
-    }
-    chromaprint_finish(ctx);
-
-    void* fprint = NULL;
-    int size = 0;
-    int ret = chromaprint_get_raw_fingerprint(ctx, &fprint, &size);
-    QByteArray fingerprint;
-    if (ret == 1) {
-        void* encoded = NULL;
-        int encoded_size = 0;
-        chromaprint_encode_fingerprint(fprint, size,
-                                       CHROMAPRINT_ALGORITHM_DEFAULT,
-                                       &encoded,
-                                       &encoded_size, 1);
-
-        fingerprint.append(reinterpret_cast<char*>(encoded), encoded_size);
-
-        chromaprint_dealloc(fprint);
-        chromaprint_dealloc(encoded);
+        Mixxx::AudioSource::size_type numFrames =
+                kFingerprintDuration * pAudioSource->getFrameRate();
+        // check that the song is actually longer then the amount of audio we use
+        if (numFrames > pAudioSource->getFrameCount()) {
+            numFrames = pAudioSource->getFrameCount();
+        }
+
+        QTime timerReadingFile;
+        timerReadingFile.start();
+
+        // Allocate a sample buffer with maximum size to avoid the
+        // implicit allocation of a temporary buffer when reducing
+        // the audio signal to stereo.
+        std::vector<Mixxx::AudioSource::sample_type> sampleBuffer(
+                math_max(numFrames * kFingerprintChannels, pAudioSource->frames2samples(numFrames)));
+
+        DEBUG_ASSERT(2 == kFingerprintChannels); // implicit assumption of the next line
+        const Mixxx::AudioSource::size_type readFrames =
+                pAudioSource->readSampleFramesStereo(
+                        numFrames, &sampleBuffer[0], sampleBuffer.size());
+        if (readFrames != numFrames) {
+            qDebug() << "oh that's embarrassing I couldn't read the track";
+            return QString();
+        }
+
+        std::vector<SAMPLE> fingerprintSamples(readFrames * kFingerprintChannels);
+        // Convert floating-point to integer
+        for (Mixxx::AudioSource::size_type i = 0; i < fingerprintSamples.size(); ++i) {
+            fingerprintSamples[i] = SAMPLE(sampleBuffer[i] * SAMPLE_MAX);
```

Comment:
here we should use the util function below convertS16ToFloat32 to have all that stuff at a common place.